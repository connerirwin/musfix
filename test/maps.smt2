; What if we allow uninterpreted functions that take no args, and are just constants?
; (declare-const m1 (Map_t Int Int) (Map_default 0))
; (declare-const name sort value)

(constraint
  (forall ((v Int) (m1 (Map_t Int Int)))
    (=> (&& (= v (Map_select m1 100)) (= m1 (Map_default 0)))
        (= v 0))))

(constraint
  (forall ((v Int) (m1 (Map_t Int Int)) (m2 (Map_t Int Int)))
    (=> (&& (&& (= v (Map_select m2 100)) (= m1 (Map_default 0)))
          (= m2 (Map_store (Map_store m1 10 1) 20 1)))
        (= v 0))))

(constraint
  (forall ((v Int) (m1 (Map_t Int Int)) (m2 (Map_t Int Int)))
    (=> (&& (&& (= v (Map_select m2 10)) (= m1 (Map_default 0)))
          (= m2 (Map_store (Map_store m1 10 1) 20 1)))
        (= v 1))))

(constraint
  (forall ((v Int) (m1 (Map_t Int Int)) (m2 (Map_t Int Int)) (m3 (Map_t Int Int)))
    (=> (&& (&& (= m1 (Map_default 0)) (= m2 (Map_store (Map_store m1 10 1) 20 1)))
          (= m3 (Map_store (Map_store m1 20 1) 10 1)))
        (= m2 m3))))

; TODO implement map_union
; (constraint
;   (forall ((v Int) (m1 (Map_t Int Int)) (m2 (Map_t Int Int)) (m3 (Map_t Int Int)) (m4 (Map_t Int Int)) (m5 (Map_t Int Int)))
;     (=> (&& (= m1 (Map_default 0)) (&& (= m2 (Map_store (Map_store m1 10 1) 20 1))
;           (&& (= m3 (Map_store (Map_store m1 20 1) 10 1))
;           (&& (= m4 (Map_store m1 10 1)) (= m5 (Map_store m1 20 1))))))
;         (= m2 (Map_union m4 m5)))))
