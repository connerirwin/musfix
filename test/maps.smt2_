(constraint
  (forall ((v int)(m1 Map_t int int))
          (=> (&& (= v (Map_select m1 100))(= m1 (Map_default 0)))
              (= v 0))))

(constraint
  (forall ((v int)(m1 Map_t int int)(m2 Map_t int int))
          (=> (&& (= v (Map_select m2 100)) (&& (= m1 (Map_default 0)) (= m2 (Map_store (Map_store m1 10 1) 20 1))))
              (= v 0))))

(constraint
  (forall ((v int)(m1 Map_t int int)(m2 Map_t int int))
          (=> (&& (= v (Map_select m2 10)) (&& (= m1 (Map_default 0)) (= m2 (Map_store (Map_store m1 10 1) 20 1))))
              (= v 1))))

; TODO what if not all of the variables are used? should we perform some sort of culling when instantiating concrete sorts?
; v int is not used
(constraint
  (forall ((v int)(m1 Map_t int int)(m2 Map_t int int)(m3 Map_t int int))
          (=> (&& (= m1 (Map_default 0)) (&& (= m2 (Map_store (Map_store m1 10 1) 20 1)) (= m3 (Map_store (Map_store m1 20 1) 10 1))))
              (= m2 m3))))

(constraint
  (forall ((v int)(m1 Map_t int int)(m2 Map_t int int)(m3 Map_t int int)(m4 Map_t int int)(m5 Map_t int int))
          (=> (&& (= m1 (Map_default 0)) (&& (= m2 (Map_store (Map_store m1 10 1) 20 1))
                  (&& (= m3 (Map_store (Map_store m1 20 1) 10 1)) (&& (= m4 (Map_store m1 10 1)) (= m5 (Map_store m1 20 1))))))
              (= m2 (Map_union m4 m5)))))
