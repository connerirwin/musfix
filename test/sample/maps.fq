
bind 1 m1 : {v : Map_t Int Int | v = Map_default 0}
bind 2 m2 : {v : Map_t Int Int | v = (Map_store (Map_store m1 10 1) 20 1) } 
bind 3 m3 : {v : Map_t Int Int | v = (Map_store (Map_store m1 20 1) 10 1) } 
bind 4 m4 : {v : Map_t Int Int | v = (Map_store m1 10 1) } 
bind 5 m5 : {v : Map_t Int Int | v = (Map_store m1 20 1) } 

constraint:
  env [ 1 ]
  lhs {v : int | v = Map_select m1 100 }
  rhs {v : int | v = 0 } 
  id 1 tag []

constraint:
  env [ 1; 2 ]
  lhs {v : int | v = Map_select m2 100 }
  rhs {v : int | v = 0 } 
  id 2 tag []

constraint:
  env [ 1; 2 ]
  lhs {v : int | v = Map_select m2 10 }
  rhs {v : int | v = 1 } 
  id 3 tag []

constraint:
  env [ 1; 2; 3 ]
  lhs {v : int | true }
  rhs {v : int | m2 = m3 } 
  id 4 tag []

constraint:
  env [ 1; 2; 3; 4; 5 ]
  lhs {v : int | true }
  rhs {v : int | m2 = Map_union m4 m4 } 
  id 5 tag []

