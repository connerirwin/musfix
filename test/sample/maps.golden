
Preparing to find greatest fixpoint...

Inputs
--------
Reading from file test/sample/maps.msmt
HornConstraint [Var IntS "v",Var (MapS IntS IntS) "m1"] (Binary Implies (Binary And (Binary Eq (Var IntS "v") (MapSel (Var (MapS IntS IntS) "m1") (IntLit 100))) (Binary Eq (Var (MapS IntS IntS) "m1") (MapLit IntS (IntLit 0)))) (Binary Eq (Var IntS "v") (IntLit 0)))
HornConstraint [Var IntS "v",Var (MapS IntS IntS) "m1",Var (MapS IntS IntS) "m2"] (Binary Implies (Binary And (Binary And (Binary Eq (Var IntS "v") (MapSel (Var (MapS IntS IntS) "m2") (IntLit 100))) (Binary Eq (Var (MapS IntS IntS) "m1") (MapLit IntS (IntLit 0)))) (Binary Eq (Var (MapS IntS IntS) "m2") (MapUpd (MapUpd (Var (MapS IntS IntS) "m1") (IntLit 10) (IntLit 1)) (IntLit 20) (IntLit 1)))) (Binary Eq (Var IntS "v") (IntLit 0)))
HornConstraint [Var IntS "v",Var (MapS IntS IntS) "m1",Var (MapS IntS IntS) "m2"] (Binary Implies (Binary And (Binary And (Binary Eq (Var IntS "v") (MapSel (Var (MapS IntS IntS) "m2") (IntLit 10))) (Binary Eq (Var (MapS IntS IntS) "m1") (MapLit IntS (IntLit 0)))) (Binary Eq (Var (MapS IntS IntS) "m2") (MapUpd (MapUpd (Var (MapS IntS IntS) "m1") (IntLit 10) (IntLit 1)) (IntLit 20) (IntLit 1)))) (Binary Eq (Var IntS "v") (IntLit 1)))
HornConstraint [Var IntS "v",Var (MapS IntS IntS) "m1",Var (MapS IntS IntS) "m2",Var (MapS IntS IntS) "m3"] (Binary Implies (Binary And (Binary And (Binary Eq (Var (MapS IntS IntS) "m1") (MapLit IntS (IntLit 0))) (Binary Eq (Var (MapS IntS IntS) "m2") (MapUpd (MapUpd (Var (MapS IntS IntS) "m1") (IntLit 10) (IntLit 1)) (IntLit 20) (IntLit 1)))) (Binary Eq (Var (MapS IntS IntS) "m3") (MapUpd (MapUpd (Var (MapS IntS IntS) "m1") (IntLit 20) (IntLit 1)) (IntLit 10) (IntLit 1)))) (Binary Eq (Var (MapS IntS IntS) "m2") (Var (MapS IntS IntS) "m3")))

QMAP
--------
fromList []

Candidates
--------
[Candidate {solution = fromList [], validConstraints = fromList [Binary Implies (Binary And (Binary Eq (Var IntS "v") (MapSel (Var (MapS IntS IntS) "m1") (IntLit 100))) (Binary Eq (Var (MapS IntS IntS) "m1") (MapLit IntS (IntLit 0)))) (Binary Eq (Var IntS "v") (IntLit 0)),Binary Implies (Binary And (Binary And (Binary Eq (Var IntS "v") (MapSel (Var (MapS IntS IntS) "m2") (IntLit 10))) (Binary Eq (Var (MapS IntS IntS) "m1") (MapLit IntS (IntLit 0)))) (Binary Eq (Var (MapS IntS IntS) "m2") (MapUpd (MapUpd (Var (MapS IntS IntS) "m1") (IntLit 10) (IntLit 1)) (IntLit 20) (IntLit 1)))) (Binary Eq (Var IntS "v") (IntLit 1)),Binary Implies (Binary And (Binary And (Binary Eq (Var IntS "v") (MapSel (Var (MapS IntS IntS) "m2") (IntLit 100))) (Binary Eq (Var (MapS IntS IntS) "m1") (MapLit IntS (IntLit 0)))) (Binary Eq (Var (MapS IntS IntS) "m2") (MapUpd (MapUpd (Var (MapS IntS IntS) "m1") (IntLit 10) (IntLit 1)) (IntLit 20) (IntLit 1)))) (Binary Eq (Var IntS "v") (IntLit 0)),Binary Implies (Binary And (Binary And (Binary Eq (Var (MapS IntS IntS) "m1") (MapLit IntS (IntLit 0))) (Binary Eq (Var (MapS IntS IntS) "m2") (MapUpd (MapUpd (Var (MapS IntS IntS) "m1") (IntLit 10) (IntLit 1)) (IntLit 20) (IntLit 1)))) (Binary Eq (Var (MapS IntS IntS) "m3") (MapUpd (MapUpd (Var (MapS IntS IntS) "m1") (IntLit 20) (IntLit 1)) (IntLit 10) (IntLit 1)))) (Binary Eq (Var (MapS IntS IntS) "m2") (Var (MapS IntS IntS) "m3"))], invalidConstraints = fromList [], label = "0"}]


Final candidates: 
0: [] (4 0)

